File: tools/test_rle_simple.c
/**
 * 测试RLE解码是否正确
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef unsigned char byte;
typedef unsigned short word;
typedef unsigned int dword;

/* RLE解码函数 - 与fd2_dat.c中的实现相同 */
int fd_decompress_rle(const byte *src, int src_size, byte *dst, int dst_width, int dst_height, int value_param) {
    int expected = dst_width * dst_height;
    int dst_idx = 0;
    int src_idx = 0;
    
    while (dst_idx < expected && src_idx < src_size) {
        byte ctrl = src[src_idx];
        src_idx++;
        
        int bit7 = (ctrl >> 7) & 1;
        int bit6 = (ctrl >> 6) & 1;
        int count = (ctrl & 0x3F) + 1;
        
        if (bit7 == 0) {
            if (bit6 == 0) {
                // FILL: 填充count个像素，值为下一个字节
                if (src_idx < src_size) {
                    byte fill_val = src[src_idx];
                    src_idx++;
                    if (value_param != -1) {
                        fill_val = (value_param + fill_val) & 0xFF;
                    }
                    for (int i = 0; i < count && dst_idx < expected; i++) {
                        dst[dst_idx] = fill_val;
                        dst_idx++;
                    }
                }
            } else {
                // SKIP: 跳过count个像素（填充0）
                dst_idx += count;
            }
        } else {
            if (bit6 == 0) {
                // COPY: 复制count个字节
                for (int i = 0; i < count && dst_idx < expected && src_idx < src_size; i++) {
                    byte val = src[src_idx];
                    src_idx++;
                    if (value_param != -1) {
                        val = (value_param + val) & 0xFF;
                    }
                    dst[dst_idx] = val;
                    dst_idx++;
                }
            } else {
                // SKIP: 跳过count个像素
                dst_idx += count;
            }
        }
    }
    
    return 0;
}

int main(int argc, char* argv[]) {
    /* 索引1的RLE数据（从调试脚本获取） */
    byte rle_data[] = {
        0x00, 0x56, 0x00, 0x00, 0x00, 0x33, 0x01, 0x00,
        0x00, 0x2e, 0x02, 0x00, 0x00, 0x1a, 0x03, 0x00,
        0x00, 0xac, 0x03, 0x00, 0x00, 0x37, 0x04, 0x00,
        0x00, 0x1e, 0x05, 0x00, 0x00, 0xa4, 0x05, 0x00,
        0x00, 0x25, 0x06, 0x00, 0x00, 0x8b, 0x06, 0x00,
        0x00, 0x12, 0x07, 0x00, 0x00, 0x72, 0x07, 0x00,
        0x00, 0xf9, 0x07, 0x00, 0x00, 0x3c, 0x08, 0x00,
        0x00, 0xc3, 0x08, 0x00, 0x00, 0x06, 0x09, 0x00,
        0x00, 0x8d, 0x09, 0x00, 0x00, 0xd0, 0x09, 0x00,
        0x00, 0x13, 0x0a, 0x00, 0x00, 0x56, 0x0a, 0x00
    };
    int rle_size = sizeof(rle_data);
    
    byte dst[576];  /* 24x24 */
    memset(dst, 0, sizeof(dst));
    
    int width = 24;
    int height = 24;
    int palette_window = 20;
    
    fd_decompress_rle(rle_data, rle_size, dst, width, height, palette_window);
    
    /* 显示解码结果 */
    printf("解码结果: 24x24\n");
    int non_zero = 0;
    for (int row = 0; row < height; row++) {
        for (int col = 0; col < width; col++) {
            int idx = row * width + col;
            byte val = dst[idx];
            if (val != 0) non_zero++;
            printf("%s", val == 0 ? "." : "#");
        }
        printf("\n");
    }
    
    printf("\n非零像素: %d / %d\n", non_zero, width * height);
    
    /* 显示前20个像素值 */
    printf("\n前20个像素值: ");
    for (int i = 0; i < 20; i++) {
        printf("%3d ", dst[i]);
    }
    printf("\n");
    
    return 0;
}

File: tools/test_rle_simple.py
import struct

with open('game/FDOTHER.DAT', 'rb') as f:
    data = f.read()

# 解析索引表
offsets = []
table_start = 6
for i in range(103):
    offset_addr = table_start + i * 4
    offset_val = struct.unpack_from('<I', data, offset_addr)[0]
    offsets.append(offset_val)

def decompress_rle_v1(src, width, height):
    """版本1: 简单的RLE解码"""
    dst = []
    idx = 0
    while len(dst) < width * height and idx < len(src):
        cmd = src[idx]
        idx += 1
        
        bit7 = (cmd >> 7) & 1
        bit6 = (cmd >> 6) & 1
        count = (cmd & 0x3F) + 1
        
        if bit7 == 0 and bit6 == 0:
            # FILL
            if idx < len(src):
                val = src[idx]
                idx += 1
                dst.extend([val] * count)
        elif bit7 == 0 and bit6 == 1:
            # SKIP
            dst.extend([0] * count)
        elif bit7 == 1 and bit6 == 0:
            # COPY
            if idx + count <= len(src):
                dst.extend(src[idx:idx+count])
                idx += count
        else:
            # SKIP
            dst.extend([0] * count)
    
    return dst

def decompress_rle_v2(src, width, height):
    """版本2: 根据MCP代码的精确实现"""
    expected = width * height
    dst = [0] * expected
    dst_pos = 0
    
    # 每行处理
    for row in range(height):
        count = width  # 当前行剩余像素
        
        while count > 0:
            if dst_pos >= len(src):
                break
            
            value = src[dst_pos]
            dst_pos += 1
            
            bit7 = (value >> 7) & 1
            bit6 = (value >> 6) & 1
            
            count_val = (value & 0x3F) + 1
            
            if bit7 == 0:
                if bit6 == 0:
                    # FILL
                    if dst_pos < len(src):
                        fill_val = src[dst_pos]
                        dst_pos += 1
                        for _ in range(count_val):
                            if count > 0:
                                # 这里需要映射到正确的dst位置
                                count -= 1
                else:
                    # SKIP
                    count -= count_val
            else:
                if bit6 == 0:
                    # COPY
                    if dst_pos + count_val <= len(src):
                        for i in range(count_val):
                            if count > 0:
                                count -= 1
                                dst_pos += 1
                    else:
                        break
                else:
                    # SKIP
                    count -= count_val
    
    return []

# 测试索引1
print('=== 测试索引1 (24x24) ===')
res1 = data[offsets[1]:offsets[2]]
w, h = struct.unpack('<HH', res1[:4])
print(f'w={w}, h={h}')

for header_size in [5, 8]:
    rle_data = res1[header_size:]
    decoded = decompress_rle_v1(rle_data, w, h)
    success = len(decoded)
    print(f'  头{header_size}字节: 解码{success}/{w*h}像素')
    
    if success == w * h:
        print('  OK! 图案:')
        for row in range(h):
            line = ''
            for col in range(w):
                v = decoded[row * w + col]
                line += '#' if v > 0 else '.'
            print(f'    {line}')

# 也测试其他tile
print('\n=== 测试其他tile资源 ===')
for idx in [11, 15, 55, 56]:
    res = data[offsets[idx]:offsets[idx+1]]
    w, h = struct.unpack('<HH', res[:4])
    
    for header_size in [5, 8]:
        if len(res) >= header_size:
            rle_data = res[header_size:]
            decoded = decompress_rle_v1(rle_data, w, h)
            status = 'OK' if len(decoded) == w * h else f'FAIL ({len(decoded)}/{w*h})'
            print(f'  索引{idx}: 头{header_size}字节 -> {status}')
